File: apps/mobile/src/api/usage.ts
// Per-turn cost lookup for the mobile client (成本呈现; 团队工资单 read side).
//
// A finished turn's spend is replayed from the cost_events ledger by message_id
// (api/routes 工资单). The LIVE turn already carries its cost in the SSE message_end
// (the fold's ProjectedTurn.cost), so this endpoint is only for RELOADED history — a
// persisted MessageDetail does not carry cost. Supplementary: callers swallow failures
// (cost must never break the chat). REST DTOs track OpenAPI via contract-rest-types.
import { apiFetch } from "@/api/client";
import type { components } from "@/types/api.generated";

type Schemas = components["schemas"];

type TurnCost = Schemas["TurnCost"];

/**
 * A turn's persisted display money in integer nano. Prefers ledger `cost.total`;
 * falls back to `estimated_cost.total` for BYOK. Returns null when neither is >0.
 *
 * Billed and estimated are separate breakdowns with separate currencies (CNY
 * curated vs USD community), so the chosen amount carries its own `currency` out
 * — the caller renders the symbol from that, never from which branch it took.
 */
export async function getMessageCostDisplay(messageId: string): Promise<{
  nano: number;
  estimated: boolean;
  currency: string;
  unpriced?: boolean;
} | null> {
  try {
    const res = await apiFetch(`/v1/messages/${messageId}/cost`);
    if (!res.ok) return null;
    const data = (await res.json()) as TurnCost;
    if ((data.cost?.total ?? 0) > 0) {
      return {
        nano: data.cost.total,
        estimated: false,
        currency: data.cost.currency,
      };
    }
    const est = data.estimated_cost?.total ?? 0;
    if (est > 0) {
      return {
        nano: est,
        estimated: true,
        currency: data.estimated_cost?.currency ?? data.cost.currency,
      };
    }
    // BYOK 三层价卡全落空：连估算都没有，但确有真实花费——上报未计价而非
    // 静默无值（拍板 2026-07-20，与桌面同口径：显式标注、金额不出数）。
    if (data.cost?.pricing_source === "unpriced") {
      return {
        nano: 0,
        estimated: false,
        currency: data.cost.currency,
        unpriced: true,
      };
    }
    return null;
  } catch {
    return null;
  }
}

// --- Account dashboard (设置·用量) ---

export type CostBreakdown = Schemas["CostBreakdown"];
export type UsageBreakdown = Schemas["UsageBreakdown"];
export type UsageWindow = Schemas["UsageWindow"];
export type QuotaStatus = Schemas["QuotaStatus"];
export type DailyCost = Schemas["DailyCost"];
export type UsageSummary = Schemas["UsageSummary"];

/** Account dashboard: today's tokens/cost, the month's cost, quota. 无 FX。 */
export async function getUsageSummary(): Promise<UsageSummary> {
  const res = await apiFetch("/v1/usage/summary");
  if (!res.ok) throw new Error(`加载用量失败 (${res.status})`);
  return (await res.json()) as UsageSummary;
}

File: apps/server/agentcore/api/schemas/usage.py
"""Cost & usage (团队工资单 + 账户仪表盘) schemas.

Money is integer nano-USD (1 USD = 1e9) everywhere — never a float. The single
display-only CNY conversion rides on ``cny_total`` (server-side via CNY_PER_USD),
so the client never re-derives money. Token fields use the ledger short keys
(matching cost_events.tokens / RunState.usage), distinct from message_end's
legacy ``*_tokens`` SSE shape.
"""

from pydantic import BaseModel


class CostBreakdown(BaseModel):
    """A run's / turn's / window's cost in integer nano-USD (canonical)."""

    input: int
    cached: int
    output: int
    total: int
    currency: str = "USD"
    # Display-only CNY value (元), converted server-side via the single
    # CNY_PER_USD rate so the client shows money without re-pricing.
    cny_total: float
    # Which price layer produced these numbers (缺省 curated 兼容旧数据).
    pricing_source: str = "curated"


class UsageBreakdown(BaseModel):
    """Token counts (cache_hit + cache_miss == input; reasoning ⊆ output)."""

    input: int
    output: int
    reasoning: int
    cache_hit: int
    cache_miss: int


class AgentCostLine(BaseModel):
    """One participant's row in the team payroll (one Run = one Agent)."""

    run_id: str
    agent_id: str | None
    role: str
    model: str
    usage: UsageBreakdown
    cost: CostBreakdown
    estimated_cost: CostBreakdown | None = None
    duration_ms: int


class TurnCost(BaseModel):
    """A turn's cost + per-Agent payroll (``GET /messages/{id}/cost``).

    Rebuilt from the ``cost_events`` ledger by message_id, so it replays a past
    turn's payroll on reload. ``agents`` is empty when the turn has no ledger
    rows (e.g. unknown / non-owned message — never leaks existence).
    """

    message_id: str
    usage: UsageBreakdown
    cost: CostBreakdown
    estimated_cost: CostBreakdown | None = None
    rounds: int
    agents: list[AgentCostLine]


class ConversationCost(BaseModel):
    """A conversation's cumulative spend (``GET /conversations/{id}/cost``)."""

    conversation_id: str
    usage: UsageBreakdown
    cost: CostBreakdown
    estimated_cost: CostBreakdown | None = None
    turns: int


class UsageWindow(BaseModel):
    """Aggregated usage over a time window (today / month)."""

    usage: UsageBreakdown
    cost: CostBreakdown
    estimated_cost: CostBreakdown | None = None
    # Distinct assistant turns in the window (the quota's「请求」proxy).
    requests: int


class QuotaStatus(BaseModel):
    """Free-tier limits (决策④); 0 = unlimited. Money is USD nano internally."""

    daily_tokens: int
    monthly_cost_nano: int
    daily_requests: int


class RoleCostLine(BaseModel):
    """One persona/role's spend over a window — the team payroll grouped by label.

    The account dashboard's product differentiator (§7.3D): multi-agent spend
    splits by ``COALESCE(persona, role)`` (调研员 / CEO / …, falling back to the
    structural captain/member bucket for legacy rows), which a single-agent
    competitor can't show. Money is integer nano-USD; the client formats ¥ from
    the summary's single ``cny_per_usd`` (no per-row re-pricing here).
    """

    role: str
    cost_total: int
    cost_estimated_total: int = 0
    # Distinct assistant turns this role took part in over the window.
    turns: int


class ModelCostLine(BaseModel):
    """One model's call-level spend over a window — admin per-model payroll.

    Aggregated from ``cost_calls`` (``GROUP BY model``), never from
    ``cost_events.model`` (that column only records the run's first call; multi-model
    runs would mis-attribute). Money is integer nano-USD; ``tokens_total`` is
    ``SUM(input + output + reasoning)``. The client formats ¥ from the summary's
    single ``cny_per_usd``.
    """

    model: str
    # LLM call rows in the window (``cost_calls`` row count).
    calls: int
    tokens_total: int
    cost_total: int
    cost_estimated_total: int = 0


class DailyCost(BaseModel):
    """One UTC day's total spend — a point in the dashboard 7-day trend (§7.3D)."""

    # ISO date (YYYY-MM-DD) of the UTC calendar day.
    date: str
    cost_total: int


class UsageSummary(BaseModel):
    """Account dashboard payload (``GET /usage/summary``).

    Also carries ``cny_per_usd`` so the client formats money from a single
    server-owned rate (it never hard-codes the FX rate).
    """

    today: UsageWindow
    month: UsageWindow
    # This month's spend split by role (团队工资单 by role), spend-desc, >0 only.
    month_by_role: list[RoleCostLine]
    # Last 7 UTC days incl today, oldest-first, zero-filled — the trend sparkline.
    recent_daily_cost: list[DailyCost]
    quota: QuotaStatus
    cny_per_usd: float
    # Billing mode (config.billing_mode). In "byok" the platform quota is dormant
    # (the turn runs on the user's own key), so the client reframes the quota meters
    # as「自带 Key 不限额」and presents cost as the user's own DeepSeek spend.
    billing_mode: str
